torch/csrc/autograd/python_function.cpp

```diff
@@ -110,15 +149,51 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
     PyTuple_SET_ITEM(pyInputs.get(), i, input);
   }
 
-  THPObjectPtr apply_fn(PyObject_GetAttrString(obj, "apply"));
-  if (!apply_fn)
-    throw_python_error();
-  THPObjectPtr r(PyObject_CallObject(apply_fn, pyInputs.get()));
+  const auto& is_variable_input = py_fn->is_variable_input;
+  const auto& input_infos = py_fn->input_info;
+  // input_info only contains info from variable inputs and should be a subset
+  TORCH_INTERNAL_ASSERT(is_variable_input.size() >= input_infos.size());
+
+  THPObjectPtr r;
+  if (!compiler.has_value() || name() == "CompiledFunctionBackward") {
+    THPObjectPtr apply_fn(PyObject_GetAttrString(obj, "apply"));
+    if (!apply_fn)
+      throw_python_error();
+    r = PyObject_CallObject(apply_fn, pyInputs.get());
+  } else {
+    // The gradients returned in the backwards should match the number of inputs
+    // to the forward, and their shapes, so we pass the fwdInputs
+    THPObjectPtr fwdInputs(
+        PyTuple_New(static_cast<Py_ssize_t>(is_variable_input.size())));
+    int offset = 0;
+    for (const auto i : c10::irange(is_variable_input.size())) {
+      if (!is_variable_input[i]) {
+        // input at i is not a variable, add an offset
+        PyTuple_SET_ITEM(fwdInputs.get(), i, Py_None);
+        offset++;
+        continue;
+      }
+
+      const auto& input_info = input_infos[i - offset];
+      PyTuple_SET_ITEM(
+          fwdInputs.get(), static_cast<Py_ssize_t>(i), THPVariable_Wrap(zeros_without_gil(input_info, _device_guard)));
```

Comment:
Is this call to zero's getting captured in the graph?  Why is this needed?

Reply:
yes it's currently captured in the graph (but I don't think we need it in there). I need a way to create tensors to bind to bwd outputs.

Their shapes are sometimes symbolic so I couldn't pass them using C API objects (https://docs.python.org/3/c-api/concrete.html). The pattern I found was the one we use for inputs which uses `THPVariable_Wrap` to convert tensors, and `zeros_without_gil` to convert from input_info/output_info to tensors

Reply:
Having them be captured is wrong, you should pass the needed sizes into the function that creates the node so that the fake outputs can be created without being included in the graph.